libs/vkd3d/command.c

```diff
@@ -14669,11 +14692,327 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetProtectedResourceSession(d3d
+
+        vk_image_subresource_layers_from_d3d12(&region.srcSubresource,
+                resolve->src_resource->format, subresource_args->SrcSubresource,
+                resolve->src_resource->desc.MipLevels, resolve->src_resource->desc.DepthOrArraySize);
+
+        vk_image_subresource_layers_from_d3d12(&region.dstSubresource,
+                resolve->dst_resource->format, subresource_args->DstSubresource,
+                resolve->dst_resource->desc.MipLevels, resolve->dst_resource->desc.DepthOrArraySize);
+
+        region.srcOffset.x = subresource_args->SrcRect.left;
+        region.srcOffset.y = subresource_args->SrcRect.top;
+        region.dstOffset.x = subresource_args->DstX;
+        region.dstOffset.y = subresource_args->DstY;
+        region.extent.width = (uint32_t)(subresource_args->SrcRect.right - subresource_args->SrcRect.left);
+        region.extent.height = (uint32_t)(subresource_args->SrcRect.bottom - subresource_args->SrcRect.top);
+        region.extent.depth = 1u;
+
+        /* Merge with existing mip region if possible. To keep things simple, only consider
+         * ordered subresource indices. */
+        merged_regions = false;
+
+        for (j = 0; j < resolve->region_count; j++)
+        {
+            src_region = &list->rtv_resolve_regions[resolve->region_index + j];
+
+            if (src_region->srcSubresource.aspectMask == region.srcSubresource.aspectMask &&
+                    src_region->srcSubresource.mipLevel == region.srcSubresource.mipLevel &&
+                    src_region->srcSubresource.baseArrayLayer + src_region->srcSubresource.layerCount == region.srcSubresource.baseArrayLayer &&
+                    src_region->dstSubresource.aspectMask == region.dstSubresource.aspectMask &&
+                    src_region->dstSubresource.mipLevel == region.dstSubresource.mipLevel &&
+                    src_region->dstSubresource.baseArrayLayer + src_region->dstSubresource.layerCount == region.dstSubresource.baseArrayLayer &&
+                    src_region->srcOffset.x == region.srcOffset.x && src_region->srcOffset.y == region.srcOffset.y &&
+                    src_region->dstOffset.x == region.dstOffset.x && src_region->dstOffset.y == region.dstOffset.y &&
+                    src_region->extent.width == region.extent.width && src_region->extent.height == region.extent.height)
+            {
+                src_region->srcSubresource.layerCount += region.srcSubresource.layerCount;
+                src_region->dstSubresource.layerCount += region.dstSubresource.layerCount;
+
+                merged_regions = true;
+                break;
+            }
+        }
+
+        if (!merged_regions)
+        {
+            vkd3d_array_reserve((void **)&list->rtv_resolve_regions, &list->rtv_resolve_region_size,
+                    list->rtv_resolve_region_count + 1, sizeof(*list->rtv_resolve_regions));
+
+            list->rtv_resolve_regions[resolve->region_index + (resolve->region_count++)] = region;
```

Comment:
Should we assert here that resolve->region_index + resolve->region_count == list->rtv_resolve_region_count? Seems like it should be true, but it's slightly confusing to not index by the list->rtv_resolve_region_count here.

Reply:
I can change this to be more clear.